pgrx/src/spi/cursor.rs

```diff
@@ -0,0 +1,118 @@
+use std::ffi::CStr;
+use std::marker::PhantomData;
+use std::ptr::NonNull;
+
+use crate::pg_sys;
+use crate::spi::SpiOkCodes;
+
+use super::client::SpiClient;
+use super::tuple::SpiTupleTable;
+use super::Error;
+
+type CursorName = String;
+
+/// An SPI Cursor from a query
+///
+/// Represents a Postgres cursor (internally, a portal), allowing to retrieve result rows a few
+/// at a time. Moreover, a cursor can be left open within a transaction, and accessed in
+/// multiple independent Spi sessions within the transaction.
+///
+/// A cursor can be created via [`SpiClient::open_cursor()`] from a query.
+/// Cursors are automatically closed on drop, unless explicitly left open using
+/// [`Self::detach_into_name()`], which returns the cursor name; cursors left open can be retrieved
+/// by name (in the same transaction) via [`SpiClient::find_cursor()`].
+///
+/// # Important notes about memory usage
+/// Result sets ([`SpiTupleTable`]s) returned by [`SpiCursor::fetch()`] will not be freed until
+/// the current Spi session is complete;
+/// this is a Pgrx limitation that might get lifted in the future.
+///
+/// In the meantime, if you're using cursors to limit memory usage, make sure to use
+/// multiple separate Spi sessions, retrieving the cursor by name.
+///
+/// # Examples
+/// ## Simple cursor
+/// ```rust,no_run
+/// use pgrx::prelude::*;
+/// # fn foo() -> spi::Result<()> {
+/// Spi::connect(|mut client| {
+///     let mut cursor = client.open_cursor("SELECT * FROM generate_series(1, 5)", None);
+///     assert_eq!(Some(1u32), cursor.fetch(1)?.get_one::<u32>()?);
+///     assert_eq!(Some(2u32), cursor.fetch(2)?.get_one::<u32>()?);
+///     assert_eq!(Some(3u32), cursor.fetch(3)?.get_one::<u32>()?);
+///     Ok::<_, pgrx::spi::Error>(())
+///     // <--- all three SpiTupleTable get freed by Spi::connect at this point
+/// })
+/// # }
+/// ```
+///
+/// ## Cursor by name
+/// ```rust,no_run
+/// use pgrx::prelude::*;
+/// # fn foo() -> spi::Result<()> {
+/// let cursor_name = Spi::connect(|mut client| {
+///     let mut cursor = client.open_cursor("SELECT * FROM generate_series(1, 5)", None);
+///     assert_eq!(Ok(Some(1u32)), cursor.fetch(1)?.get_one::<u32>());
+///     Ok::<_, spi::Error>(cursor.detach_into_name()) // <-- cursor gets dropped here
+///     // <--- first SpiTupleTable gets freed by Spi::connect at this point
+/// })?;
+/// Spi::connect(|mut client| {
+///     let mut cursor = client.find_cursor(&cursor_name)?;
+///     assert_eq!(Ok(Some(2u32)), cursor.fetch(1)?.get_one::<u32>());
+///     drop(cursor); // <-- cursor gets dropped here
+///     // ... more code ...
+///     Ok(())
+///     // <--- second SpiTupleTable gets freed by Spi::connect at this point
+/// })
+/// # }
+/// ```
+pub struct SpiCursor<'client> {
+    pub(crate) ptr: NonNull<pg_sys::PortalData>,
+    pub(crate) __marker: PhantomData<&'client SpiClient<'client>>,
+}
```

Comment:
It's not entirely obvious to me why this is the lifetime marker instead of e.g. only allowing access to SpiCursor via borrowing it from an SpiClient.

Reply:
In general, `pgrx::spi` has a problem where the historical implementations were either slightly clumsy, which is to be expected, or too clever by half. This is a bad combination, as it became "idiomatic", as it were, to follow the existing codebase style and also construct more peculiar implementations that rely on either unnecessary or over-clever idioms, instead of trying to use a much more direct and "boring" approach. We'd still have to fix things either way, however, so it's not worth dwelling on.